pyresample/utils.py

```diff
@@ -298,6 +298,110 @@ def get_area_def(area_id, area_name, proj_id, proj4_args, x_size, y_size,
+        else:
+            raise ValueError('The source raster is not gereferenced, '
+                             'please provide the value of proj_dict')
+
+        if proj_id is None:
+            proj_id = proj.split('"')[1]
+
+    area_def = AreaDefinition(area_id, name, proj_id, proj_dict,
+                              dataset.RasterXSize, dataset.RasterYSize,
+                              area_extent)
+    return area_def
+
+
+def _get_area_def_from_rasterio(dataset, area_id, name, proj_id=None,
+                                proj_dict=None):
+    from pyresample.geometry import AreaDefinition
+
+    a, b, c, d, e, f, _, _, _ = dataset.transform
+    if not (b == d == 0):
+        raise ValueError('Rotated rasters are not supported at this time.')
+
+    if proj_dict is None:
+        crs = dataset.crs
+        if crs is not None:
+            proj_dict = dataset.crs.to_dict()
+        else:
+            raise ValueError('The source raster is not gereferenced, '
+                             'please provide the value of proj_dict')
+
+        if proj_id is None:
+            proj_id = crs.wkt.split('"')[1]
+
+    area_def = AreaDefinition(area_id, name, proj_id, proj_dict,
+                              dataset.width, dataset.height, dataset.bounds)
+    return area_def
+
+
+def get_area_def_from_raster(source, area_id=None, name=None, proj_id=None,
+                             proj_dict=None):
+    """Construct AreaDefinition object from raster
+
+    Parameters
+    ----------
+    source : str, Dataset, DatasetReader or DatasetWriter
+        A file name. Also it can be ``osgeo.gdal.Dataset``,
+        ``rasterio.io.DatasetReader`` or ``rasterio.io.DatasetWriter``
+    area_id : str, optional
+        ID of area
+    name : str, optional
+        Name of area
+    proj_id : str, optional
+        ID of projection
+    proj_dict : dict, optional
+        PROJ.4 parameters
+
+    Returns
+    -------
+    area_def : object
+        AreaDefinition object
+    """
+    source_class_name = source.__class__.__name__
+    if source_class_name in ('DatasetReader', 'DatasetWriter'):
+        return _get_area_def_from_rasterio(source, area_id, name, proj_dict)
+    elif source_class_name == 'Dataset':
+        return _get_area_def_from_gdal(source, area_id, name,
+                                       proj_id, proj_dict)
+    else:
+        try:
+            import rasterio
+            with rasterio.open(source) as src:
+                return _get_area_def_from_rasterio(src, area_id, name,
+                                                   proj_id, proj_dict)
+        except ImportError:
+            try:
+                from osgeo import gdal
+                src = gdal.Open(source)
+                return _get_area_def_from_gdal(src, area_id, name,
+                                               proj_id, proj_dict)
+            except ImportError:
+                raise ImportError('Either rasterio or gdal must be available')
+
```

Comment:
I think you could shorten this whole function and reduce the amount of duplicate code if you:

1. `if isinstance(src, str): # try import and open` first then did the class checks.
2. Take advantage of the 120 line length for the `_get_area_def_...` calls.
3. I'd prefer if it was an `isinstance` check for the class type, but I understand that might make the code ugly. Maybe at least change it to:

```python
if source.__class__.__name__ == 'Dataset':
    # do gdal stuff
else:
    # do rasterio stuff
```

With the import stuff you could consider (go with what you feel):

```python
try:
    import rasterio
except ImportError:
    rasterio = None
    from osgeo import gdal

if isinstance(source, str) and rasterio:
    source = rasterio.open(source)
elif isinstance(source, str):
    source = gdal.Open(source)
...
```

Just a thought.
4. Should we explicitly close the source files? iirc gdal doesn't have a close method and relies on `__del__` cleanup.

Reply:
Thanks for quick reply! I modified code according to your suggestion. Anything else?